pkg/capabilities/triggers/cron_trigger.go

```diff
@@ -0,0 +1,287 @@
+}
+
+func (cts *CronTriggerService) process(triggerId string, schedule cron.Schedule) {
+	cts.mu.Lock()
+	defer cts.mu.Unlock()
+
+	now := time.Now()
+	scheduledExecutionTime := schedule.Next(now)
+	// if scheduledExecutionTime is before now, get the next one
+	// this can be due to time sync / sleep
+	for {
+		if scheduledExecutionTime.Before(now) {
+			scheduledExecutionTime = schedule.Next(now.Add(time.Second))
+		} else {
+			break
+		}
+	}
+
+	// timestamp in ns precision, to show the discrepancy between the scheduled time and the actual time
+	// we need scheduled time to generate a deterministic triggerEventId
+	timestampNs := time.Now().UTC().UnixNano()
+	// format to ISO 8601
+	timestampMsIso8601 := formatUnixNanoToISO8601(timestampNs)
+
+	scheduledExecutionTimeFormatted := scheduledExecutionTime.Format(time.RFC3339)
+	hash := sha256.Sum256([]byte(scheduledExecutionTimeFormatted))
+	triggerEventId := hex.EncodeToString(hash[:])
+
+	cts.lggr.Debugw("process", "scheduledExecTime", scheduledExecutionTime, "actualExecTime", timestampMsIso8601, "triggerEventId", triggerEventId)
+
+	capabilityResponse, err := wrapCronTriggerEvent(triggerEventId, timestampMsIso8601, scheduledExecutionTimeFormatted)
+	if err != nil {
+		cts.lggr.Errorw("error wrapping trigger event", "err", err)
+	}
+
+	select {
+	case cts.jobs[triggerId].ch <- capabilityResponse:
+	default:
+		cts.lggr.Errorw("channel full, dropping event", "eventID", triggerEventId, "triggerID", triggerID, "capabilityResponse", capabilityResponse)
+	}
+}
+
+func (cts *CronTriggerService) UnregisterTrigger(ctx context.Context, req capabilities.CapabilityRequest) error {
+	cts.mu.Lock()
+	defer cts.mu.Unlock()
+
+	config, err := cts.ValidateConfig(req.Config)
+	if err != nil {
+		return err
+	}
+
+	triggerID := getTriggerID(config.TriggerID, req.Metadata.WorkflowID)
+
+	if _, ok := cts.jobs[triggerID]; !ok {
+		return fmt.Errorf("triggerId %s not found", triggerID)
+	}
+
+	trigger, ok := cts.jobs[triggerID]
+	jobId := trigger.job.ID()
+
+	if ok {
+		cts.scheduler.RemoveJob(jobId)
+		close(trigger.ch)
+	}
+	delete(cts.jobs, triggerID)
+	cts.lggr.Debugw("UnregisterTrigger", "triggerId", triggerID, "jobId", jobId)
+	return nil
+}
+
+// Start the service.
+func (cts *CronTriggerService) Start(ctx context.Context) error {
+	cts.wg.Add(1)
+
+	if cts.scheduler == nil {
+		return errors.New("no scheduler initialized")
+	}
+	cts.scheduler.Start()
+
+	cts.lggr.Info(cts.Name() + " started")
+
+	// block until ready to shut down
+	go func() {
+		defer cts.wg.Done()
+		<-cts.stopCh
+	}()
+
+	return nil
+}
+
+// Close stops the Service.
+// Invariants: After this call the Service cannot be started
+// again, new Service will need to be built to do so.
+func (cts *CronTriggerService) Close() error {
+	err := cts.scheduler.Shutdown()
+	if err != nil {
+		return fmt.Errorf("scheduler shutdown encountered a problem: %s", err)
+	}
+
+	close(cts.stopCh)
+
+	cts.wg.Wait()
```

Comment:
This two bits seem unnecessary: there are no goroutines that are managed directly by this service, and all we do is start the scheduler in Start(); it feels like all we should do here is stop it.